Documentation/network/multicast.rst

```diff
@@ -0,0 +1,109 @@
+   Kubernetes cluster. Please see :ref:`k8s_quick_install` for more
+   information. If unsure, run ``cilium status`` and validate that Cilium is up
+   and running. This guide also assumes Cilium is configured with vxlan mode,
+   which is required when using multicast capability.
+
+Enable Multicast Feature
+========================
+
+This feature requires a Cilium ``v1.16.0`` or later, and is disabled by default.
+When enabling the multicast, Cilium must be configured with vxlan mode.
+
+Multicast support can be enabled by updating ``cilium-config`` ConfigMap as following:
+
+.. code-block:: bash
+
+   $ cilium config set multicast-enabled true
+   ✨ Patching ConfigMap cilium-config with multicast-enabled=true...
+   ♻️  Restarted Cilium pods
+
+
+Configure Multicast and Subscriber IPs
+======================================
+
+To use multicast with Cilium, we need to configure multicast group IP addresses and
+subscriber list based on the application requirement by ``cilium-dbg`` command.
+After the setting, multicast subscriber pods can send out IGMP join and multicast
+sender pods can start sending multicast stream.
+
+As an example, the following procedure uses ``239.25.0.1`` multicast group address.
+Note that multicast IP addresses and subscribers need to be set on each ``cilium-agent``.
+
+Run the commands to get all CiliumNode IP addresses to be set as multicast subscribers.
+
+.. code-block:: bash
+
+   $ kubectl get ciliumnodes.cilium.io
+   NAME                 CILIUMINTERNALIP   INTERNALIP   AGE
+   kind-control-plane   10.244.0.72        172.19.0.2   16m
+   kind-worker          10.244.1.86        172.19.0.3   16m
+
+To set multicast IP address, run the following commands to enable multicast BPF maps in each ``cilium-agent``
+
+.. code-block:: bash
+
+   ### add multicast IP address
+   $ cilium-dbg bpf multicast group add 239.255.0.1
+
+   ### check multicast IP address
+   $ cilium-dbg bpf multicast group list
+   Group Address
+   239.255.0.1
+
+Then, we can set the subscriber IP addresses as followings in each ``cilium-agent``.
+
+.. code-block:: bash
+
+   ### cilium-agent on kind-control-plane
+   $ cilium-dbg bpf multicast subscriber add 239.255.0.1 10.244.1.86
+   $ cilium-dbg bpf multicast subscriber list all
+   Group           Subscriber      Type
+   239.255.0.1     10.244.1.86     Remote Node
+
+   ### cilium-agent on kind-worker
+   $ cilium-dbg bpf multicast subscriber add 239.255.0.1 10.244.0.72
+
+.. note::
+
+   This multicast subscriber IP addresses are other CiliumNode IP addresses than your own one.
+
+When you want to remove multicast IP addresses and subscriber list, run the following commands in the ``cilium-agent``.
+
+.. code-block:: bash
+
+   $ cilium-dbg bpf multicast group delete
+   $ cilium-dbg bpf multicast subscriber delete 239.255.0.1 10.244.0.72
+
+
+Limitations
+===========
+
+* The operation needs to be done on each CiliumNode that uses multicast feature.
```

Comment:
(Not related to this PR) currently the whole operation requires user to log in each cilium-agent and execute the appropriate command, this could be burden for user experience. that would be probably nice to set and configure via `cilium-cli` at once with user arguments. if this is the right way to take, we can take care of this cli enhancement as well.

Reply:
Yes, that would be nice improvement. There will need to be some logic in cilium-cli to automatically detect nodes and configure each one. I think only these CLIs are required
- cilium multicast group <list|add|delete> <group-ip>
- cilium multicast subscriber list <group-ip>

Reply:
in addition to that, there would be the case that we want to enable multicast group within specific `ciliumnode`s only? having the multicast packets in all ciliumnodes scaled up to hundreds, this would be not efficient? of course that application needs to be controlled to deploy the specific ciliumnodes where multicast group is enabled.

if this is the case, maybe we need to be able to specify the ciliumnodes to enable multicast? what do you think?

one of our developer would like to make contribution for this. so you can leave this task to us, we will create another issue for this cli extension development.

Reply:
> one of our developer would like to make contribution for this. so you can leave this task to us, we will create another issue for this cli extension development.

👆 CC: @joestringer @lambdanis

Reply:
> if this is the case, maybe we need to be able to specify the ciliumnodes to enable multicast? what do you think?

Yes, doing cluster wide blanket configuration with lot of nodes is going to cause overhead in terms of packet replication. In which case, I think we are going into the territory of writing a control plane, which should probably not sit inside cilium-cli. There are bunch of moving parts, which are being taken care outside of cilium-cli. You may be setting node labels, and pod affinity rules etc. 

My initial thought is to keep cilium-cli cluster wide in terms of configuring multicast subscribers across all cluster nodes. This would help in configuring small scale clusters. For larger clusters, I would recommend having dedicated operator/control plane to manage individual cilium agents.

Reply:
sounds good to me 👍  if needs the extension, we can always do so. we will start with your initial idea, which is concrete.

> I think we are going into the territory of writing a control plane
> You may be setting node labels, and pod affinity rules etc

originally i was thinking that, user would configure specific multicast IP for framework-A with only selected cilium nodes, and framework-B and so on. (different multicast IP addresses and different cilium nodes mapping) and that would be complicated to configure control-plane. i may be missing something, anyway i agree with you. we can keep the discussion on further enhancement before implementation.

Reply:
resolving this thread for now.